### PR TITLE
Add `theme` parameter to `get_stylesheet`

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
There are a few different places where we repeat this pattern:
```python
from napari.qt import get_stylesheet
from napari.utils.theme import get_theme, template

raw_stylesheet = get_stylesheet()
theme = get_theme(theme_name)
obj.setStyleSheet(template(raw_stylesheet, **theme))
```
Not only does that force lots of imports everywhere, but it's repetitive, and doesn't take advantage of stylesheet caching (since the template needs to be reapplied each time).  This PR lets `get_stylesheet` optionally apply the theme templating:

```python
from napari.qt import get_stylesheet

qt_obj.setStyleSheet(get_stylesheet(theme_name))
```

It also removes the manual `QStatusBar` theming, which is no longer necessary with proper widget lineages.  The _only_ two blockers to full stylesheet inheritance now are the `canvas` object (which does not use stylesheets and needs to manually set `bgcolor`), and the `napari-console`, due to the fact that `qtconsole` also doesn't use the native stylesheet inheritance, but rather requires you to manually set it.

@sofroniewn, this PR will break napari-console, and requires the following change there:
```python

# qt_console.py
    def _update_theme(self, event=None):
        """Update the napari GUI theme."""
        from napari.utils.theme import get_theme
        from napari.qt import get_stylesheet

        theme = get_theme(self.viewer.theme)
		# qtconsole unfortunately won't inherit the parent stylesheet
        self.style_sheet = get_stylesheet(self.viewer.theme)
        self.syntax_style = theme["syntax_style"]
        bracket_color = QColor(*str_to_rgb(theme["highlight"]))
        self._bracket_matcher.format.setBackground(bracket_color)
``` 

It would be easy enough to also accomplish #2103 here too, we just need to decide on the API. (for instance, should it be `viewer.canvas_background_override` ... or should it be somewhere only in qt land, like `viewer.window.qt_viewer.canvas.background_override`  (my preference would be the latter, since we don't really "have" a public canvas concept at the moment

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
This looks great @tlambert03, I very much like the changes 

> @sofroniewn, this PR will break napari-console, and requires the following change there:

This confuses me, the PR looks backwards compatible if no theme arg is passed, why does this change need to be make?

It would be nice if we could make an update to napari-console that allowed it to work both with the old syntax in 0.4.5 and this new syntax as we could then release `0.0.3` and avoid compatibility issues. Is that the case with the code snippet you provided? If so happy to get that added and an `0.0.3` release made before we merge this PR.

> `viewer.window.qt_viewer.canvas.background_override` (my preference would be the latter, since we don't really "have" a public canvas concept at the moment

Yeah I think for now I'd much rather have this live on the qt objects until we have a public `viewer.canvas`. We can do that in a follow up PR though if we want.

**Comment by tlambert03** [comment]
> the PR looks backwards compatible if no theme arg is passed, why does this change need to be make?

It's not a syntax issue... the bit that makes it backwards incompatible was removing `raw_stylesheet` as a class attribute from `qt_viewer`.  I can undo that if you want to change it slowly.  However, I can't see any reason why that should be there (other than to support napari-console [here](https://github.com/napari/napari-console/blob/3313a710ad671532727c6d2f4f8d255afdc6bb05/napari_console/qt_console.py#L156)).  our global style template isn't related to `qt_viewer` any more than any other qt object.  Moreover `raw_stylesheet` is already the cached result of calling `get_stylesheet`, so it's doubly strange that it gets saved on the qt_viewer.

**Comment by sofroniewn** [comment]
> It's not a syntax issue... the bit that makes it backwards incompatible was removing raw_stylesheet as a class attribute from qt_viewer. I can undo that if you want to change it slowly. However, I can't see any reason why that should be there (other than to support napari-console here). our global style template isn't related to qt_viewer any more than any other qt object. Moreover raw_stylesheet is already the cached result of calling get_stylesheet, so it's doubly strange that it gets saved on the qt_viewer.

Ah right I get it. It totally makes sense for it to go, and I'm happy for it to go as quickly as it can without us ending up with napari-console or napari breaking.

I'm trying to think about how to coordinate the merge of this PR, the update and release of the next version of napari-console, and the release of the next version of napari.

I guess we could update napari-console right now to have an `if hasattr` that looks for the `raw_stylesheet`. If there it does what it used to do, if not it then does the new syntax. This would be napari-console 0.0.3.

We could then merge this PR, but I'd probably still add a deprecation warning to remove in 0.4.7 or 0.4.8 if possible as @Czaki might need this attribute.

Once napari 0.4.6 gets released we could go back to napari-console, remove that `if hasattr` and release `0.0.4`.

How does this sound @tlambert03?

**Comment by tlambert03** [comment]
Sounds good!

**Comment by sofroniewn** [comment]
> Sounds good!

PR to napari-console made here https://github.com/napari/napari-console/pull/11

**Comment by tlambert03** [comment]
I added a `__getattr__` with a deprecation warning to both window and qt_viewer.  I also removed all changes to `.canvas`, which can be done elsewhere.

File: napari/_qt/qt_main_window.py
Comment (Czaki) [inline napari/_qt/qt_main_window.py:164]: please. Left at least 2 versions between removal.

Comment (tlambert03) [reply napari/_qt/qt_main_window.py:164]: we can do that.  but just out of curiosity what does that buy you?  You can access the exact same value from `qt.get_stylesheet()`... why do you want to be able to access it here across multiple deprecated versions?

File: napari/_qt/qt_resources/__init__.py
Comment (Czaki) [inline napari/_qt/qt_resources/__init__.py:140]: It looks like it changes current napari behavior. Currently `get_theme` and `register_theme` allow overwriting some theme parameters. I use it to overwrite the canvas background to have a black background with the `light` theme.

File: napari/_qt/qt_main_window.py
Comment (Czaki) [reply napari/_qt/qt_main_window.py:164]: Personally, I do not use it. But I do not like the idea of 1 version deprecation.

Comment (tlambert03) [reply napari/_qt/qt_main_window.py:164]: I'm 99% sure that you're the only person who would be using it! :joy:

File: napari/_qt/qt_resources/__init__.py
Comment (tlambert03) [reply napari/_qt/qt_resources/__init__.py:140]: The one thing this changes (due to caching) is that if someone were to modify a theme after setting that theme, those changes would not be reflected.  While that worked, I don't think we ever intended to "support" themes as mutable objects.  `register_theme` will allow you to create a new modified theme that will work here.  We understand you'd like to have the canvas background be settable apart from the theme, (#2103 and [mentioned above](https://github.com/napari/napari/pull/2263#issue-572951065)).  That's a good feature, very possible and unrelated to this.  We can enable that in another PR.

Comment (Czaki) [reply napari/_qt/qt_resources/__init__.py:140]: I agree. But #2103 stuck with feedback that its needs to be discussed. If this PR will be merged and #2103 not then I lost functionality. 

And still, If both PR was merged then my current code may be broken. There may be missed one version to keep both versions working.

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py:164]: i'm then fine leaving this at a 1 version deprecation - this is inside out Qt folder and we're still in alpha.

File: napari/_qt/qt_resources/__init__.py
Comment (sofroniewn) [reply napari/_qt/qt_resources/__init__.py:140]: > While that worked, I don't think we ever intended to "support" themes as mutable objects.

Yes, this was not an intended piece of functionality, but an unintended side effect.

I'd like to merge this PR now as it is improving the stylesheet getting. We can talk more about the canvas override in #2103. I think the best approach there might be a private attribute that you use until we have a public canvas object.

**Comment by sofroniewn** [comment]
Ok, changes look good @tlambert03.

> I added a __getattr__ with a deprecation warning to both window and qt_viewer.

Thanks for the deprecation warning. As I said above, one version is fine for me. 

> I also removed all changes to .canvas, which can be done elsewhere.

I liked these changes, but agree they can be in their own PR or a follow up.

I will get napari-console 0.0.3 released and then merge this

**Review by sofroniewn — APPROVED** [review]

File: napari/_qt/qt_resources/__init__.py
Comment (Czaki) [reply napari/_qt/qt_resources/__init__.py:140]: The first discussion was in #2074 and @jni Told me that I could assume that theme is mutable
https://github.com/napari/napari/pull/2074#discussion_r551792658

Comment (sofroniewn) [reply napari/_qt/qt_resources/__init__.py:140]: I think what @jni intended in that comment is to force registering of the new theme and not leverage the mutability of the dict, see also this https://github.com/napari/napari/pull/2013#discussion_r545142762. Personally I'd prefer us not to use mutable objects here, but I lost that discussion in #2013. We'll leave this open until @jni has a chance to respond too

Comment (Czaki) [reply napari/_qt/qt_resources/__init__.py:140]: I still worry that there could be release with this PR and without possibility to set canvas color.

Comment (tlambert03) [reply napari/_qt/qt_resources/__init__.py:140]: no need to worry: https://github.com/napari/napari/pull/2270

Comment (sofroniewn) [reply napari/_qt/qt_resources/__init__.py:140]: #2270 looks great, I just approved it. I think we now have a clear path forward to merge this PR followed by that one. I'll still wait in case anyone else has additional concerns, but I'm very happy with this approach

**Comment by sofroniewn** [comment]
OK #2270 got an approval from @jni, I will merge this now. Thanks all!!

**Pull request merged**
